### PR TITLE
storage: limit total unique revisions in unsynced watcher event list

### PR DESCRIPTION
**Pull request by heyitsanthony**

### CODE REVIEW

File: storage/watchable_store.go [hunk 1]
```diff
@@ -34,11 +34,63 @@ const (
...
+}
+
 type (
 	watcherSetByKey map[string]watcherSet
 	watcherSet      map[*watcher]struct{}
+	watcherBatch    map[*watcher]*eventBatch
 )
 
+func (eb *eventBatch) add(ev storagepb.Event) {
+	if eb.moreRev != 0 {
```
Comment (xiang90) [inline storage/watchable_store.go]: i feel this is confusing. can we do `eb.revs > watchBatchMaxRevs` instead?

File: storage/watchable_store.go [hunk 2]
```diff
@@ -34,11 +34,63 @@ const (
...
+		return
+	}
+
+	// revision accounting
+	ebRev := eb.evs[len(eb.evs)-1].Kv.ModRevision
+	evRev := ev.Kv.ModRevision
+	if evRev > ebRev {
+		eb.revs++
+		if eb.revs > watchBatchMaxRevs {
+			eb.moreRev = evRev
```
Comment (xiang90) [inline storage/watchable_store.go]: if we do https://github.com/coreos/etcd/pull/4556/files#r53365921, we can always update moreRev.

File: storage/watchable_store.go [hunk 3]
```diff
@@ -393,16 +449,19 @@ func kvsToEvents(revs, vals [][]byte, wsk watcherSetByKey, pfxs map[string]struc
...
+	we := newWatcherBatch(s.synced, evs)
 	for _, wm := range s.synced {
 		for w := range wm {
-			es, ok := we[w]
+			eb, ok := we[w]
 			if !ok {
 				continue
 			}
+			if eb.revs != 1 {
+				panic("multi rev notify")
```
Comment (xiang90) [inline storage/watchable_store.go]: unexpected multiple revisions in notification.

File: storage/watchable_store.go [hunk 4]
```diff
@@ -34,11 +34,63 @@ const (
 	chanBufLen = 1024
 )
 
+var (
+	// watchBatchMaxRevs is the maximum distinct revisions that
+	// may be sent to an unsynced watcher at a time.
+	watchBatchMaxRevs = 1000
```
Comment (xiang90) [inline storage/watchable_store.go]: add comment: we declare it as a ver instead of const for testability?

File: storage/watchable_store.go (hunk 2 above)
Comment (heyitsanthony) [inline storage/watchable_store.go]: but it'll fall through and append one event for the overflow revision

**Comment by xiang90** [comment]
LGTM.

File: storage/watchable_store.go (hunk 2 above)
Comment (xiang90) [inline storage/watchable_store.go]: you are right.

Comment (xiang90) [inline storage/watchable_store.go]: probably update this after the append.

File: storage/watchable_store.go (hunk 1 above)
Comment (heyitsanthony) [inline storage/watchable_store.go]: OK

File: storage/watchable_store.go (hunk 3 above)
Comment (heyitsanthony) [inline storage/watchable_store.go]: OK

File: storage/watchable_store.go (hunk 4 above)
Comment (heyitsanthony) [inline storage/watchable_store.go]: OK

File: storage/watchable_store.go (hunk 2 above)
Comment (heyitsanthony) [inline storage/watchable_store.go]: not sure if that's much better, plus it's an extra condition per event

**Comment by heyitsanthony** [comment]
ok merging on ci greenlight

**Pull request merged**